AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLDocument/SyncEnabledGraphQLDocument.swift

```diff
@@ -0,0 +1,82 @@
+//
+// Copyright 2018-2020 Amazon.com,
+// Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Amplify
+import Foundation
+
+public class SyncEnabledGraphQLDocument<Document: GraphQLDocument>: GraphQLDocument {
+
+    private let graphQLDocument: Document
+    private let version: Int?
+
+    public init(_ graphqQLDocument: Document, version: Int? = nil) {
+        self.graphQLDocument = graphqQLDocument
+        self.version = version
+    }
+
+    public var documentType: GraphQLDocumentType {
+        graphQLDocument.documentType
+    }
+
+    public var name: String {
+        graphQLDocument.name
+    }
+
+    public var modelType: Model.Type {
+        graphQLDocument.modelType
+    }
+
+    public var stringValue: String {
+        let selectionSetString = selectionSetFields.map { $0.toString() }.joined(separator: "\n    ")
+        if let inputTypes = inputTypes, let inputParameters = inputParameters {
+            return """
+            \(documentType) \(name.pascalCased())(\(inputTypes)) {
+              \(name)(\(inputParameters)) {
+                \(selectionSetString)
+              }
+            }
+            """
+        }
+
+        return """
+        \(documentType) \(name.pascalCased()) {
+          \(name) {
+            \(selectionSetString)
+          }
+        }
+        """
+    }
+
+    public var inputTypes: String? {
+        graphQLDocument.inputTypes
+    }
+
+    public var inputParameters: String? {
+        graphQLDocument.inputParameters
+    }
+
+    public var variables: [String: Any] {
+        var variables = graphQLDocument.variables
+
+        if let version = version, var input = variables["input"] as? [String: Any] {
```

Comment:
What, if any, guarantees do we provide around consistency of the schema demanding a `_version` and us supplying it? Is it possible for the Input document generated by the CLI to want a _version but the variable to not supply it? What happens in that case, does the service generate an error?

Reply:
the cases which the developer needs to be aware of is:
1. Create Mutation, Queries, and Subscribe do not need version. Create mutation does allow a version but will fail with internal error response from the service. There shouldn't be a use case where create mutation needs a version, but i think it might just be how the transformer code was written that all mutations including Create result in a PutItem call to DynamoDB. (similar to why a ConditionInput exists on a create mutation)
2. Update Mutation and Delete Mutation contains optional version in the variables. if not passed in, will bump existing version. If passed in, I'm assuming will do conflict resolution on service side

i tried representing this in client side logic but then decided to create a more flexible decorator which simply takes in an optional version, thus leaving it up to the caller to know what they are doing.

- currently the caller needs to know what they are doing to supply or not supply a version on the underlying graphQLDocument instance they are decorating.
- yes, it is possible . currently CLI transforms the schema and wants a _version for create/update/delete mutation. create should never be created with a version. update/delete is optional. 
- only in the create mutation scenario, have i seen an error generated by the service

Reply:
Sounds like we still have some sharp edges on this. If not already in our backlog, please add an issue to resolve this with CLI team/transformer, and let's do what we can to reduce cases where a customer might be put into a place where they have to understand the inner workings of our conflict resolution strategies.